vm/src/exceptions.rs

```diff
@@ -253,6 +241,39 @@ fn write_traceback_entry<W: Write>(
     Ok(())
 }
 
+fn write_exception_recursive<W: Write>(
+    output: &mut W,
+    vm: &VirtualMachine,
+    exc: &PyBaseExceptionRef,
+    seen: &mut HashSet<usize>,
+) -> Result<(), W::Error> {
+    // This function should not be called directly,
+    // use `wite_exception` as a public interface.
+    // It is similar to `print_exception_recursive` from `CPython`.
+    seen.insert(exc.as_object().get_id());
+    if let Some(cause) = exc.cause() {
+        // This can be a special case: `raise e from e`,
+        // we just ignore it and treat like `raise e` without any extra steps.
+        if !seen.contains(&cause.as_object().get_id()) {
+            write_exception_recursive(output, vm, &cause, seen)?;
+            writeln!(
+                output,
+                "\nThe above exception was the direct cause of the following exception:\n"
+            )?;
+        } else {
+            seen.insert(cause.as_object().get_id());
+        }
+    } else if let Some(context) = exc.context() {
+        write_exception_recursive(output, vm, &context, seen)?;
```

Comment:
Pretty sure this *should* spin if `e.__context__ = e`. Checking the set again before recursing, as done with `cause`, should do the trick.

Reply:
Yes, you are right. I was trying to somehow reproduce it, but I didn't think about direct manipulation 🙈 😆 
Fixed!

Reply:
Yeah, Python allows some pretty wild and head-scratching manipulation. I wasn't even sure if a setter was available for `__context__` in RustPython.